File: Design-Patterns-Using-JavaScript/SOLID/LiskovSubstitutionPrinciple2.js
// Animal Class
class Animal {
    constructor(name) {
        this.name = name;
    }
}

// Bird Class
class Bird extends Animal {
    constructor(name) {
        super(name);
        this.flightHeight = 0;
    }

    fly() {
        throw new Error("You have to implement the method fly!");
    }
}

// Sparrow Class
class Sparrow extends Bird {
    constructor(name) {
        super(name);
    }

    fly() {
        console.log("The sparrow is fluttering its wings.");
        this.flightHeight = 100;
    }
}

// Penguin Class
class Penguin extends Bird {
    constructor(name) {
        super(name);
    }

    fly() {
        console.log("The penguin cannot fly.");
    }

    slide() {
        console.log("The penguin is sliding on its belly!");
    }

    swim() {
        console.log("The penguin is swimming in the water!");
    }
}

// Dodo Class
class Dodo extends Bird {
    constructor(name) {
        super(name);
    }

    fly() {
        console.log("The dodo is extinct and cannot fly.");
    }
}

// Test Function
function test(bird) {
    bird.fly();
    if (bird.flightHeight > 0) {
        console.log("Bird is flying at a positive height.");
    } else {
        console.log("Error: fly() method called; flight height is still zero.");
    }
}

// Client code
const sparrow = new Sparrow("Sparrow");
test(sparrow);

const penguin = new Penguin("Penguin");
test(penguin);

const dodo = new Dodo("Dodo");
test(dodo);

/*
The sparrow is fluttering its wings.
Bird is flying at a positive height.
The penguin cannot fly.
Error: fly() method called; flight height is still zero.
The dodo is extinct and cannot fly.
Error: fly() method called; flight height is still zero.
*/

File: Design-Patterns-Using-Cpp/SOLID/LiskovSubstitutionPrinciple2.cpp
#include <iostream>
#include <string>

class Animal {
protected:
    std::string name;

public:
    Animal(const std::string& name) : name(name) {}
};

class Bird : public Animal {
protected:
    int flightHeight;

public:
    Bird(const std::string& name) : Animal(name), flightHeight(0) {}

    virtual void fly() = 0;

    virtual int getFlightHeight(){
        return flightHeight;
    }
};

class Sparrow : public Bird {
public:
    Sparrow(const std::string& name) : Bird(name) {}

    void fly() override {
        std::cout << "The sparrow is fluttering its wings." << std::endl;
        flightHeight = 100;
    }
};

class Penguin : public Bird {
public:
    Penguin(const std::string& name) : Bird(name) {}

    void fly() override {
        std::cout << "The penguin cannot fly." << std::endl;
    }

    void slide() {
        std::cout << "The penguin is sliding on its belly!" << std::endl;
    }

    void swim() {
        std::cout << "The penguin is swimming in the water!" << std::endl;
    }
};

class Dodo : public Bird {
public:
    Dodo(const std::string& name) : Bird(name) {}

    void fly() override {
        std::cout << "The dodo is extinct and cannot fly." << std::endl;
    }
};

void test(Bird* bird) {
    bird->fly();
    if (bird->getFlightHeight() > 0) {
        std::cout << "Bird is flying at a positive height." << std::endl;
    } else {
        std::cout << "Error: fly() method called; flight height is still zero." << std::endl;
    }
}

// Client code
int main() {
    Sparrow sparrow("Sparrow");
    test(&sparrow);

    Penguin penguin("Penguin");
    test(&penguin);

    Dodo dodo("Dodo");
    test(&dodo);

    return 0;
}


/*
The sparrow is fluttering its wings.
Bird is flying at a positive height.
The penguin cannot fly.
Error: fly() method called; flight height is still zero.
The dodo is extinct and cannot fly.
Error: fly() method called; flight height is still zero.
*/